From e586e09d483ba0a6a5df6c50baadc76c98f31afc Mon Sep 17 00:00:00 2001
From: Jonas Karlman <jonas@kwiboo.se>
Date: Mon, 15 Dec 2025 18:17:24 +0000
Subject: [PATCH 07/10] CVideoBufferDRMPRIMEFFmpeg: Map HW AVFrame to
 AV_PIX_FMT_DRM_PRIME

Rebased onto kodi c0d95ea0, which moved IsValid() from
CVideoBufferDRMPRIMEFFmpeg up into CVideoBufferDRMPRIME. The override is
added back in the derived class and defers to the base for the
descriptor check instead of repeating it.
---
 .../Buffers/VideoBufferDRMPRIME.cpp           | 43 +++++++++++++++++++++++
 .../VideoPlayer/Buffers/VideoBufferDRMPRIME.h | 12 +++++++-----
 2 files changed, 50 insertions(+), 5 deletions(-)

diff --git a/xbmc/cores/VideoPlayer/Buffers/VideoBufferDRMPRIME.cpp b/xbmc/cores/VideoPlayer/Buffers/VideoBufferDRMPRIME.cpp
--- a/xbmc/cores/VideoPlayer/Buffers/VideoBufferDRMPRIME.cpp
+++ b/xbmc/cores/VideoPlayer/Buffers/VideoBufferDRMPRIME.cpp
@@ -85,6 +85,49 @@
   av_frame_unref(m_pFrame);
 }
 
+AVDRMFrameDescriptor* CVideoBufferDRMPRIMEFFmpeg::GetDescriptor() const
+{
+  if (m_pMapFrame)
+    return reinterpret_cast<AVDRMFrameDescriptor*>(m_pMapFrame->data[0]);
+
+  return reinterpret_cast<AVDRMFrameDescriptor*>(m_pFrame->data[0]);
+}
+
+bool CVideoBufferDRMPRIMEFFmpeg::RequiresMapping() const
+{
+  return m_pFrame->format != AV_PIX_FMT_DRM_PRIME && m_pFrame->hw_frames_ctx;
+}
+
+bool CVideoBufferDRMPRIMEFFmpeg::IsValid() const
+{
+  if (RequiresMapping())
+    return true;
+
+  return CVideoBufferDRMPRIME::IsValid();
+}
+
+bool CVideoBufferDRMPRIMEFFmpeg::AcquireDescriptor()
+{
+  if (RequiresMapping())
+  {
+    m_pMapFrame = av_frame_alloc();
+    m_pMapFrame->format = AV_PIX_FMT_DRM_PRIME;
+
+    if (av_hwframe_map(m_pMapFrame, m_pFrame, AV_HWFRAME_MAP_READ))
+    {
+      av_frame_free(&m_pMapFrame);
+      return false;
+    }
+  }
+
+  return true;
+}
+
+void CVideoBufferDRMPRIMEFFmpeg::ReleaseDescriptor()
+{
+  av_frame_free(&m_pMapFrame);
+}
+
 CVideoBufferPoolDRMPRIMEFFmpeg::~CVideoBufferPoolDRMPRIMEFFmpeg()
 {
   for (auto buf : m_all)
diff --git a/xbmc/cores/VideoPlayer/Buffers/VideoBufferDRMPRIME.h b/xbmc/cores/VideoPlayer/Buffers/VideoBufferDRMPRIME.h
--- a/xbmc/cores/VideoPlayer/Buffers/VideoBufferDRMPRIME.h
+++ b/xbmc/cores/VideoPlayer/Buffers/VideoBufferDRMPRIME.h
@@ -55,13 +55,16 @@
   void SetRef(AVFrame* frame);
   void Unref();
 
-  AVDRMFrameDescriptor* GetDescriptor() const override
-  {
-    return reinterpret_cast<AVDRMFrameDescriptor*>(m_pFrame->data[0]);
-  }
+  AVDRMFrameDescriptor* GetDescriptor() const override;
+  bool IsValid() const override;
+  bool AcquireDescriptor() override;
+  void ReleaseDescriptor() override;
 
 protected:
+  bool RequiresMapping() const;
+
   AVFrame* m_pFrame = nullptr;
+  AVFrame* m_pMapFrame = nullptr;
 };
 
 class CVideoBufferPoolDRMPRIMEFFmpeg : public IVideoBufferPool
-- 
2.34.1
